### ISSUE TITLE
ActiveSupport tests hang with Ruby Head

### ISSUE THREAD

**Issue by utilum** [rails/rails#37676]
I've been seeing this for a while, and could now reduce it to the combination of `LoadInterlockAwareMonitorTest#test_entering_with_blocking` 
with either
`ConstantLookupTest#test_does_not_swallow_exception_on_no_method_error` 
or 
`ConstantLookupTest#test_does_not_swallow_exception_on_no_name_error_within_constant`.

### Steps to reproduce

```
$ ruby -v
ruby 2.7.0dev (2019-11-10T05:56:38Z master 4570284ce1) [x86_64-linux]

$ TESTOPTS="--seed=0 --name=/ConstantLookupTest#test_does_not_swallow_exception_on_no_method_error\|LoadInterlockAwareMonitorTest#test_entering_with_blocking/" bundle exec rake

ruby 2.7.0dev (2019-11-10T05:56:38Z master 4570284ce1) [x86_64-linux]
/home/u/.rbenv/versions/2.7.0-dev/bin/ruby -w -I"lib:test" -I"/home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/rake-13.0.0/lib" "/home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/rake-13.0.0/lib/rake/rake_test_loader.rb" "test/actionable_error_test.rb" "test/array_inquirer_test.rb" "test/autoload_test.rb" "test/benchmarkable_test.rb" "test/broadcast_logger_test.rb" "test/cache/cache_entry_test.rb" "test/cache/cache_key_test.rb" "test/cache/cache_store_logger_test.rb" "test/cache/cache_store_namespace_test.rb" "test/cache/cache_store_setting_test.rb" "test/cache/local_cache_middleware_test.rb" "test/cache/stores/file_store_test.rb" "test/cache/stores/mem_cache_store_test.rb" "test/cache/stores/memory_store_test.rb" "test/cache/stores/null_store_test.rb" "test/cache/stores/redis_cache_store_test.rb" "test/callback_inheritance_test.rb" "test/callbacks_test.rb" "test/class_cache_test.rb" "test/clean_backtrace_test.rb" "test/clean_logger_test.rb" "test/concern_test.rb" "test/concurrency/load_interlock_aware_monitor_test.rb" "test/configurable_test.rb" "test/core_ext/array/access_test.rb" "test/core_ext/array/conversions_test.rb" "test/core_ext/array/extract_options_test.rb" "test/core_ext/array/extract_test.rb" "test/core_ext/array/grouping_test.rb" "test/core_ext/array/prepend_append_test.rb" "test/core_ext/array/wrap_test.rb" "test/core_ext/bigdecimal_test.rb" "test/core_ext/class/attribute_test.rb" "test/core_ext/class_test.rb" "test/core_ext/date_and_time_compatibility_test.rb" "test/core_ext/date_ext_test.rb" "test/core_ext/date_time_ext_test.rb" "test/core_ext/digest/uuid_test.rb" "test/core_ext/duration_test.rb" "test/core_ext/enumerable_test.rb" "test/core_ext/file_test.rb" "test/core_ext/hash/transform_values_test.rb" "test/core_ext/hash_ext_test.rb" "test/core_ext/integer_ext_test.rb" "test/core_ext/kernel/concern_test.rb" "test/core_ext/kernel_test.rb" "test/core_ext/load_error_test.rb" "test/core_ext/marshal_test.rb" "test/core_ext/module/anonymous_test.rb" "test/core_ext/module/attr_internal_test.rb" "test/core_ext/module/attribute_accessor_per_thread_test.rb" "test/core_ext/module/attribute_accessor_test.rb" "test/core_ext/module/attribute_aliasing_test.rb" "test/core_ext/module/concerning_test.rb" "test/core_ext/module/introspection_test.rb" "test/core_ext/module/remove_method_test.rb" "test/core_ext/module_test.rb" "test/core_ext/name_error_test.rb" "test/core_ext/numeric_ext_test.rb" "test/core_ext/object/acts_like_test.rb" "test/core_ext/object/blank_test.rb" "test/core_ext/object/deep_dup_test.rb" "test/core_ext/object/duplicable_test.rb" "test/core_ext/object/inclusion_test.rb" "test/core_ext/object/instance_variables_test.rb" "test/core_ext/object/json_cherry_pick_test.rb" "test/core_ext/object/json_gem_encoding_test.rb" "test/core_ext/object/to_param_test.rb" "test/core_ext/object/to_query_test.rb" "test/core_ext/object/try_test.rb" "test/core_ext/range_ext_test.rb" "test/core_ext/regexp_ext_test.rb" "test/core_ext/secure_random_test.rb" "test/core_ext/string_ext_test.rb" "test/core_ext/time_ext_test.rb" "test/core_ext/time_with_zone_test.rb" "test/core_ext/uri_ext_test.rb" "test/current_attributes_test.rb" "test/dependencies_test.rb" "test/deprecation/method_wrappers_test.rb" "test/deprecation/proxy_wrappers_test.rb" "test/deprecation_test.rb" "test/descendants_tracker_with_autoloading_test.rb" "test/descendants_tracker_without_autoloading_test.rb" "test/digest_test.rb" "test/encrypted_configuration_test.rb" "test/encrypted_file_test.rb" "test/evented_file_update_checker_test.rb" "test/executor_test.rb" "test/file_update_checker_test.rb" "test/fork_tracker_test.rb" "test/gzip_test.rb" "test/hash_with_indifferent_access_test.rb" "test/i18n_test.rb" "test/inflector_test.rb" "test/json/decoding_test.rb" "test/json/encoding_test.rb" "test/key_generator_test.rb" "test/lazy_load_hooks_test.rb" "test/log_subscriber_test.rb" "test/logger_test.rb" "test/message_encryptor_test.rb" "test/message_verifier_test.rb" "test/messages/rotation_configuration_test.rb" "test/multibyte_chars_test.rb" "test/multibyte_conformance_test.rb" "test/multibyte_grapheme_break_conformance_test.rb" "test/multibyte_normalization_conformance_test.rb" "test/multibyte_proxy_test.rb" "test/notifications/evented_notification_test.rb" "test/notifications/instrumenter_test.rb" "test/notifications_test.rb" "test/number_helper_i18n_test.rb" "test/number_helper_test.rb" "test/option_merger_test.rb" "test/ordered_hash_test.rb" "test/ordered_options_test.rb" "test/parameter_filter_test.rb" "test/reloader_test.rb" "test/rescuable_test.rb" "test/safe_buffer_test.rb" "test/secure_compare_rotator_test.rb" "test/security_utils_test.rb" "test/share_lock_test.rb" "test/silence_logger_test.rb" "test/string_inquirer_test.rb" "test/subscriber_test.rb" "test/tagged_logging_test.rb" "test/test_case_test.rb" "test/testing/after_teardown_test.rb" "test/testing/constant_lookup_test.rb" "test/testing/file_fixtures_test.rb" "test/testing/method_call_assertions_test.rb" "test/time_travel_test.rb" "test/time_zone_test.rb" "test/transliterate_test.rb" "test/xml_mini/jdom_engine_test.rb" "test/xml_mini/libxml_engine_test.rb" "test/xml_mini/libxmlsax_engine_test.rb" "test/xml_mini/nokogiri_engine_test.rb" "test/xml_mini/nokogirisax_engine_test.rb" "test/xml_mini/rexml_engine_test.rb" "test/xml_mini/xml_mini_engine_test.rb" "test/xml_mini_test.rb" "test/zeitwerk_inflector_test.rb" --seed=0 --name=/ConstantLookupTest#test_does_not_swallow_exception_on_no_method_error\|LoadInterlockAwareMonitorTest#test_entering_with_blocking/
/home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/tzinfo-1.2.5/lib/tzinfo/ruby_core_support.rb:142: warning: The last argument is used as the keyword parameter
/home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/tzinfo-1.2.5/lib/tzinfo/ruby_core_support.rb:142: warning: The last argument is used as the keyword parameter
WARNING: Nokogiri was built against LibXML version 2.9.9, but has dynamically loaded 2.9.4
Run options: --seed=0 "--name=/ConstantLookupTest#test_does_not_swallow_exception_on_no_method_error|LoadInterlockAwareMonitorTest#test_entering_with_blocking/"

# Running:

E

Error:
ActiveSupport::Concurrency::LoadInterlockAwareMonitorTest#test_entering_with_blocking:
fatal: No live threads left. Deadlock?
6 threads, 6 sleeps current:0x000055c980c573e0 main thread:0x000055c97e652480
* #<Thread:0x000055c97e67ef38 sleep_forever>
   rb_thread_t:0x000055c97e652480 native:0x00007f3e9e923700 int:0
   /home/u/code/rails/activesupport/test/concurrency/load_interlock_aware_monitor_test.rb:47:in `join'
   /home/u/code/rails/activesupport/test/concurrency/load_interlock_aware_monitor_test.rb:47:in `test_entering_with_blocking'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest/test.rb:98:in `block (3 levels) in run'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest/test.rb:195:in `capture_exceptions'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest/test.rb:95:in `block (2 levels) in run'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest.rb:268:in `time_it'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest/test.rb:94:in `block in run'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest.rb:363:in `on_signal'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest/test.rb:211:in `with_info_handler'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest/test.rb:93:in `run'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest.rb:1014:in `run_one_method'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest.rb:337:in `run_one_method'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest.rb:324:in `block (2 levels) in run'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest.rb:323:in `each'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest.rb:323:in `block in run'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest.rb:363:in `on_signal'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest.rb:350:in `with_info_handler'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest.rb:322:in `run'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest.rb:162:in `block in __run'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest.rb:162:in `map'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest.rb:162:in `__run'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest.rb:139:in `run'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest.rb:66:in `block in autorun'
* #<Thread:0x000055c97fb23560@/home/u/.rbenv/versions/2.7.0-dev/lib/ruby/2.7.0/drb/drb.rb:1273 sleep_forever>
   rb_thread_t:0x000055c97fab1150 native:0x00007f3e979da700 int:0
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/2.7.0/monitor.rb:108:in `sleep'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/2.7.0/monitor.rb:108:in `wait'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/2.7.0/monitor.rb:108:in `wait_for_cond'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/2.7.0/monitor.rb:108:in `wait'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/2.7.0/monitor.rb:125:in `wait_until'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/2.7.0/drb/drb.rb:1252:in `block in _execute'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/2.7.0/monitor.rb:202:in `synchronize'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/2.7.0/monitor.rb:202:in `mon_synchronize'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/2.7.0/drb/drb.rb:1251:in `_execute'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/2.7.0/drb/drb.rb:1276:in `block in make_pool'
* #<Thread:0x000055c97f08e618@/home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest/parallel.rb:28 sleep_forever>
   rb_thread_t:0x000055c980f6bba0 native:0x00007f3e973c3700 int:0
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest/parallel.rb:30:in `pop'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest/parallel.rb:30:in `block (2 levels) in start'
* #<Thread:0x000055c97f08d740@/home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest/parallel.rb:28 sleep_forever>
   rb_thread_t:0x000055c980f86240 native:0x00007f3e8d59c700 int:0
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest/parallel.rb:30:in `pop'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/gems/2.7.0/gems/minitest-5.12.2/lib/minitest/parallel.rb:30:in `block (2 levels) in start'
* #<Thread:0x000055c980b1c458@/home/u/code/rails/activesupport/test/concurrency/load_interlock_aware_monitor_test.rb:25 sleep_forever>
   rb_thread_t:0x000055c980c573e0 native:0x00007f3e8d39b700 int:0
    depended by: tb_thread_id:0x000055c97e652480
   /home/u/code/rails/activesupport/test/concurrency/load_interlock_aware_monitor_test.rb:30:in `synchronize'
   /home/u/code/rails/activesupport/test/concurrency/load_interlock_aware_monitor_test.rb:30:in `block (2 levels) in test_entering_with_blocking'
   /home/u/code/rails/activesupport/lib/active_support/dependencies/interlock.rb:42:in `block in running'
   /home/u/code/rails/activesupport/lib/active_support/concurrency/share_lock.rb:162:in `sharing'
   /home/u/code/rails/activesupport/lib/active_support/dependencies/interlock.rb:41:in `running'
   /home/u/code/rails/activesupport/test/concurrency/load_interlock_aware_monitor_test.rb:26:in `block in test_entering_with_blocking'
* #<Thread:0x000055c980b1c368@/home/u/code/rails/activesupport/test/concurrency/load_interlock_aware_monitor_test.rb:36 sleep_forever>
   rb_thread_t:0x000055c980bc6000 native:0x00007f3e8d19a700 int:0
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/2.7.0/monitor.rb:108:in `sleep'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/2.7.0/monitor.rb:108:in `wait'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/2.7.0/monitor.rb:108:in `wait_for_cond'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/2.7.0/monitor.rb:108:in `wait'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/2.7.0/monitor.rb:116:in `wait_while'
   /home/u/code/rails/activesupport/lib/active_support/concurrency/share_lock.rb:220:in `wait_for'
   /home/u/code/rails/activesupport/lib/active_support/concurrency/share_lock.rb:83:in `block (2 levels) in start_exclusive'
   /home/u/code/rails/activesupport/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
   /home/u/code/rails/activesupport/lib/active_support/concurrency/share_lock.rb:82:in `block in start_exclusive'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/2.7.0/monitor.rb:202:in `synchronize'
   /home/u/.rbenv/versions/2.7.0-dev/lib/ruby/2.7.0/monitor.rb:202:in `mon_synchronize'
   /home/u/code/rails/activesupport/lib/active_support/concurrency/share_lock.rb:77:in `start_exclusive'
   /home/u/code/rails/activesupport/lib/active_support/concurrency/share_lock.rb:149:in `exclusive'
   /home/u/code/rails/activesupport/lib/active_support/dependencies/interlock.rb:13:in `loading'
   /home/u/code/rails/activesupport/test/concurrency/load_interlock_aware_monitor_test.rb:41:in `block (2 levels) in test_entering_with_blocking'
   /home/u/code/rails/activesupport/test/concurrency/load_interlock_aware_monitor_test.rb:37:in `synchronize'
   /home/u/code/rails/activesupport/test/concurrency/load_interlock_aware_monitor_test.rb:37:in `block in test_entering_with_blocking'

    /home/u/code/rails/activesupport/test/concurrency/load_interlock_aware_monitor_test.rb:47:in `join'
    /home/u/code/rails/activesupport/test/concurrency/load_interlock_aware_monitor_test.rb:47:in `test_entering_with_blocking'

rails test home/u/code/rails/activesupport/test/concurrency/load_interlock_aware_monitor_test.rb:18

 
```

**Comment by yahonda** [comment]
According to `git bisect`, this error reproduces since https://github.com/ruby/ruby/commit/caac5f777ae288b5982708b8690e712e1cae0cf6 by https://github.com/ruby/ruby/pull/2576

```
$ git bisect bad caac5f777ae288b5982708b8690e712e1cae0cf6
caac5f777ae288b5982708b8690e712e1cae0cf6 is the first bad commit
commit caac5f777ae288b5982708b8690e712e1cae0cf6
Author: Koichi Sasada <ko1@atdot.net>
Date:   Sun Oct 20 04:52:20 2019 +0900

    make monitor.so for performance. (#2576)

    Recent monitor.rb has performance problem because of interrupt
    handlers. 'Monitor#synchronize' is frequently used primitive
    so the performance of this method is important.

    This patch rewrite 'monitor.rb' with 'monitor.so' (C-extension)
    and make it faster. See [Feature #16255] for details.

    Monitor class objects are normal object which include MonitorMixin.
    This patch introduce a Monitor class which is implemented on C
    and MonitorMixin uses Monitor object as re-entrant (recursive)
    Mutex. This technique improve performance because we don't need
    to care atomicity and we don't need accesses to instance variables
    any more on Monitor class.

 NEWS                         |   3 +
 ext/monitor/depend           |  13 ++
 ext/monitor/extconf.rb       |   2 +
 ext/monitor/lib/monitor.rb   | 287 ++++++++++++++++++++++++++++++++++++++
 ext/monitor/monitor.c        | 189 +++++++++++++++++++++++++
 lib/monitor.rb               | 323 -------------------------------------------
 test/monitor/test_monitor.rb |  16 +--
 7 files changed, 502 insertions(+), 331 deletions(-)
 create mode 100644 ext/monitor/depend
 create mode 100644 ext/monitor/extconf.rb
 create mode 100644 ext/monitor/lib/monitor.rb
 create mode 100644 ext/monitor/monitor.c
 delete mode 100644 lib/monitor.rb
$
```

**Comment by yahonda** [comment]
Opened an issue at https://bugs.ruby-lang.org/issues/16398

**Comment by yahonda** [comment]
Note: this code itself has been introduced by https://github.com/rails/rails/pull/31035

**Comment by yahonda** [comment]
This issue has been addressed by #38069
